File: app/src/main/java/com/moisesborges/tvaddict/shows/ShowsPresenter.java
package com.moisesborges.tvaddict.shows;

import android.support.annotation.NonNull;

import com.moisesborges.tvaddict.data.ShowsRepository;
import com.moisesborges.tvaddict.models.Show;
import com.moisesborges.tvaddict.mvp.BasePresenter;
import com.moisesborges.tvaddict.mvp.RxJavaConfig;

import javax.inject.Inject;

import io.reactivex.disposables.Disposable;

/**
 * Created by Moisés on 12/04/2017.
 */

public class ShowsPresenter extends BasePresenter<ShowsView> {

    private final ShowsRepository mShowsRepository;

    @Inject
    public ShowsPresenter(@NonNull ShowsRepository showsRepository,
                          @NonNull RxJavaConfig rxJavaConfig) {
        super(rxJavaConfig);
        mShowsRepository = showsRepository;
    }

    public void loadShows() {
        checkView();

        getView().displayProgress(true);

        Disposable disposable = mShowsRepository.getShows()
                .subscribeOn(getRxJavaConfig().ioScheduler())
                .observeOn(getRxJavaConfig().androidScheduler())
                .subscribe(shows -> {
                            getView().displayProgress(false);
                            getView().displayTvShows(shows);

                        }, throwable -> {
                            getView().displayProgress(false);
                            getView().displayError();
                        }
                );

        addDisposable(disposable);
    }


    public void openShowDetails(@NonNull Show show) {
        checkView();
        getView().navigateToShowDetails(show);
    }
}

File: app/src/main/java/com/moisesborges/tvaddict/shows/ShowsPresenter.kt
package com.moisesborges.tvaddict.shows

import com.moisesborges.tvaddict.data.ShowsRepository
import com.moisesborges.tvaddict.models.Show
import com.moisesborges.tvaddict.mvp.BasePresenter
import com.moisesborges.tvaddict.mvp.RxJavaConfig
import com.raizlabs.android.dbflow.kotlinextensions.save
import io.reactivex.functions.Consumer
import timber.log.Timber

import javax.inject.Inject

/**
 * Created by Moisés on 12/04/2017.
 */

class ShowsPresenter
@Inject constructor(private val showsRepository: ShowsRepository,
                    rxJavaConfig: RxJavaConfig) : BasePresenter<ShowsView>(rxJavaConfig) {

    fun loadShows() {
        checkView()

        view.displayProgress(true)

        val disposable = showsRepository.loadShows(0)
                .map(mapToUnfollowingListOfShows())
                .compose(applySchedulersToSingle())
                .subscribe({ shows ->
                    view.displayProgress(false)
                    view.displayTvShows(shows)
                    view.setPage(0)

                }
                ) { throwable ->
                    view.displayProgress(false)
                    view.displayError()
                }

        addDisposable(disposable)
    }

    private fun mapToUnfollowingListOfShows(): (List<Show>) -> List<Show> {
        return { showsFromWeb ->
            val savedShows = showsRepository.getSavedShows().blockingGet()
            showsFromWeb.filter { savedShows.count { savedShow -> savedShow.id == it.id } == 0 }
        }
    }


    fun openShowDetails(show: Show) {
        checkView()
        view.navigateToShowDetails(show)
    }

    fun loadMoreShows(currentPage: Int) {
        checkView()

        val nextPage = currentPage + 1
        val disposable = showsRepository.loadShows(nextPage)
                .map(mapToUnfollowingListOfShows())
                .compose(applySchedulersToSingle())
                .subscribe({ shows ->
                    view.displayMoreTvShows(shows)
                    view.setPage(nextPage)
                }, { error -> view.displayError() })

        addDisposable(disposable)
    }

    fun followShow(show: Show) {
        checkView()

        val disposable = showsRepository.saveFullShowInfo(show)
                .compose(applySchedulersToSingle())
                .subscribe({ savedShow ->
                    view.hideShow(savedShow)
                    view.displayFollowingShowMessage(savedShow)
                }, {error -> Timber.e(error)})

        addDisposable(disposable)
    }
}
